### PR TITLE
feat(storage): implement storage::GrpcClient::DeleteResumableUpload

### PR DESCRIPTION
**Pull request by remyabel**
Fixes: #5910

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5941)

### CODE REVIEW

File: google/cloud/storage/internal/object_requests.h [hunk 1]
```diff
@@ -197,6 +197,15 @@ class DeleteObjectRequest
                                   IfMetagenerationNotMatch, UserProject> {
  public:
   using GenericObjectRequest::GenericObjectRequest;
+
+  std::string const& upload_id() const { return upload_id_; }
```
Comment (coryan) [inline google/cloud/storage/internal/object_requests.h:201]: This change seems unrelated, this is a `DeleteObjectRequest`, not a `DeleteResumableUploadRequest`, it seems the proto APIs make no distinction, but we do here.

File: google/cloud/storage/internal/grpc_client.cc [hunk 2]
```diff
@@ -378,8 +378,20 @@ GrpcClient::RestoreResumableSession(std::string const& upload_url) {
...
+    DeleteResumableUploadRequest const& request) {
+  grpc::ClientContext context;
+  auto upload_session_params =
+      DecodeGrpcResumableUploadSessionUrl(request.upload_session_url());
+  google::protobuf::Empty response;
+  DeleteObjectRequest delete_request;
+  delete_request.set_bucket_name(upload_session_params->bucket_name);
+  delete_request.set_object_name(upload_session_params->object_name);
+  delete_request.set_upload_id(upload_session_params->upload_id);
+  auto proto_request = ToProto(delete_request);
```
Comment (coryan) [inline google/cloud/storage/internal/grpc_client.cc:390]: I think we want a new `google::storage::v1::DeleteObject ToProto(DeleteResumableUploadRequest const&)` function, then the body of this function gets simpler.

File: google/cloud/storage/tests/object_integration_test.cc [hunk 3]
```diff
@@ -898,8 +898,6 @@ TEST_F(ObjectIntegrationTest, DeleteAccessControlFailure) {
 }
 
 TEST_F(ObjectIntegrationTest, DeleteResumableUpload) {
-  // TODO(#5910) - DeleteResumableUpload doesn't work with gRPC yet.
-  if (UsingGrpc()) GTEST_SKIP();
```
Comment (coryan) [inline google/cloud/storage/tests/object_integration_test.cc:902]: Note that some of these RPCs are only implemented in the emulator for gRPC, production only implements the basic upload and download operations at the moment.

**Review by coryan — CHANGES_REQUESTED** [review]
See below, I think we can simplify these changes.

File: google/cloud/storage/internal/object_requests.h (hunk 1 above)
Comment (remyabel) [reply google/cloud/storage/internal/object_requests.h]: Fixed.

File: google/cloud/storage/internal/grpc_client.cc (hunk 2 above)
Comment (remyabel) [reply google/cloud/storage/internal/grpc_client.cc]: Fixed.

File: google/cloud/storage/tests/object_integration_test.cc (hunk 3 above)
Comment (remyabel) [reply google/cloud/storage/tests/object_integration_test.cc:902]: Fixed.

**Comment by coryan** [comment]
Don't worry about the build breaks, that was an internal snafu on the key refresh.

**Comment by coryan** [comment]
Consider rebasing before pushing any changes. The version numbers changed and that invalidates all the build caches.

**Comment by remyabel** [comment]
I'll do that but what other changes did you need me to make?

**Review by coryan — CHANGES_REQUESTED** [review]
Doh! I wrote this review this morning I never sent it.

File: google/cloud/storage/internal/grpc_client.cc [hunk 4]
```diff
@@ -1806,6 +1812,17 @@ google::storage::v1::QueryWriteStatusRequest GrpcClient::ToProto(
   return r;
 }
 
+google::storage::v1::DeleteObjectRequest GrpcClient::ToProto(
+    DeleteResumableUploadRequest const& request) {
+  auto upload_session_params =
+      DecodeGrpcResumableUploadSessionUrl(request.upload_session_url());
+  google::storage::v1::DeleteObjectRequest r;
+  r.set_bucket(upload_session_params->bucket_name);
```
[the remark this answers is not in the document]
Comment (remyabel) [reply google/cloud/storage/internal/grpc_client.cc]: `DecodeGrpcResumableUploadSessionUrl()` seems fine, the extra fields are needed for other cases just not this one. Fixed.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 1 of 4 files at r1, 3 of 4 files at r2, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5941#-:-MUk61GB7Mf_x6rTE1WK:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved
